File: LeetCode/0799_ChampagneTower/lc0799.cpp
#include <cassert>
#include <vector>

class Solution {
public:
  double champagneTower(int poured, int query_row, int query_glass) {
    // initially all the champagne gets poured directly into the top glass
    std::vector<double> flow_previous_row(1, poured);

    for (std::size_t num_row = 1; num_row <= query_row; num_row++) {
      // there are i+1 glasses in ith row (since rows are 0-indexed), all initially empty.
      std::vector<double> flow_current_row(num_row + 1, 0);

      // for each glass in parent row, calculate how much of champagne spills equally to both children:
      for (std::size_t g = 0; g < num_row; g++) {
        // 1 glass of champagne stays in a top glass, the excess spills equally to two glasses below:
        double excess_flow = 0.5 * (flow_previous_row[g] - 1.0);
        if (excess_flow > 0.0) {
          flow_current_row[g] += excess_flow;
          flow_current_row[g+1] += excess_flow;
        }
      }

      // Just before exiting this loop, current flow becomes previous flow.
      flow_previous_row = flow_current_row;
    }
    return std::min(1.0, flow_previous_row[query_glass]);
  };
};

int main() {
  Solution s;
  assert(s.champagneTower(1, 1, 1) == 0.0);
  assert(s.champagneTower(2, 1, 1) == 0.5);
  assert(s.champagneTower(100000009, 33, 17) == 1.0);
}

File: LeetCode/0799/golang/lc0799.go
package lc0799

// Runtime complexity: O(query_row^2)
// Auxiliary space: O(query_row)
// Subjective level: medium
// Solved on: 2026-02-14
func champagneTower(poured int, query_row int, query_glass int) float64 {
	// Pour all the `poured` champagne directly into the top glass (0, 0):
	flow_previous_row := []float64{float64(poured)}
	for curr_row := 1; curr_row <= query_row; curr_row++ {
		// As the tower's rows are 0-indexed, there are (n+1) glasses in nth row:
		flow_curr_row := make([]float64, curr_row+1)

		// For each of the glasses in the previous row, calculate how much of the champagne
		// spills equally to both descendant glasses. Remember that one glass of champagne
		// remains in the "parent" (top) glass:
		for g := range curr_row {
			excess_flow := 0.5 * (flow_previous_row[g] - 1.0)
			if excess_flow > 0.0 {
				flow_curr_row[g] += excess_flow
				flow_curr_row[g+1] += excess_flow
			}
		}
		// At the end of each iteration, the "current" flow becomes the "previous" flow
		flow_previous_row = flow_curr_row
	}

	// Return what's at the queried row's glass; cap the result at at most 1.0 (i.e. glass being full).
	return min(flow_previous_row[query_glass], 1.0)
}
